### PR TITLE
Declare some types

### PR DESCRIPTION
**Pull request by lcobucci**
Since now we're require PHP 7.2 we can start to add `object`

### CODE REVIEW

**Comment by lcobucci** [comment]
I've tried to minimise the impact so I could already use this in https://github.com/doctrine/doctrine2/pull/7054

File: lib/Doctrine/ORM/Sequencing/Generator.php [hunk 1]
```diff
@@ -11,11 +11,9 @@
     /**
      * Generates an identifier for an entity.
      *
-     * @param object $entity
-     *
-     * @return \Generator
+     * @return string|int
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Sequencing/Generator.php]: What happened here? :|

File: lib/Doctrine/ORM/Sequencing/SequenceGenerator.php [hunk 2]
```diff
@@ -70,41 +64,31 @@ public function generate(EntityManagerInterface $em, $entity)
...
-    public function getCurrentMaxValue()
+    public function getCurrentMaxValue() : ?int
     {
         return $this->maxValue;
     }
 
     /**
      * Gets the next value that will be returned by generate().
-     *
-     * @return int
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Sequencing/SequenceGenerator.php]: Was this previously wrong?

File: lib/Doctrine/ORM/Sequencing/SequenceGenerator.php [hunk 3]
```diff
@@ -70,41 +64,31 @@ public function generate(EntityManagerInterface $em, $entity)
...
+                'sequenceName'   => $this->sequenceName,
             ]
         );
     }
 
-    /**
-     * @param string $serialized
-     */
-    public function unserialize($serialized)
+    public function unserialize($serialized) : void
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Sequencing/SequenceGenerator.php]: Can a `string` type be added here? Or does it collide with `Serializable`?

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/Sequencing/SequenceGenerator.php (hunk 3 above)
Comment (lcobucci) [reply lib/Doctrine/ORM/Sequencing/SequenceGenerator.php]: It collides with `Serializable`

File: lib/Doctrine/ORM/Sequencing/SequenceGenerator.php (hunk 2 above)
Comment (lcobucci) [reply lib/Doctrine/ORM/Sequencing/SequenceGenerator.php]: yeap...

Comment (lcobucci) [reply lib/Doctrine/ORM/Sequencing/SequenceGenerator.php]: I mean, it's is not needed with the type hint =)

File: lib/Doctrine/ORM/Sequencing/Generator.php (hunk 1 above)
Comment (lcobucci) [reply lib/Doctrine/ORM/Sequencing/Generator.php]: It was just wrong before, you can see based on the implementation...

File: lib/Doctrine/ORM/Sequencing/SequenceGenerator.php (hunk 2 above)
Comment (Ocramius) [reply lib/Doctrine/ORM/Sequencing/SequenceGenerator.php]: The hint added nullability

Comment (Ocramius) [reply lib/Doctrine/ORM/Sequencing/SequenceGenerator.php]: Hmm, no, I think my comment was misplaced

File: lib/Doctrine/ORM/Sequencing/SequenceGenerator.php (hunk 3 above)
Comment (lcobucci) [reply lib/Doctrine/ORM/Sequencing/SequenceGenerator.php]: I'll put the docblock back, my intention was to add the type declaration

Comment (lcobucci) [reply lib/Doctrine/ORM/Sequencing/SequenceGenerator.php]: done

**Review by Ocramius — APPROVED** [review]

**Comment by Ocramius** [comment]
:ship:

**Comment by Majkl578** [comment]
@Ocramius Hmm so now documentation of type changes is not required anymore? I am confused.

**Comment by Ocramius** [comment]
@Majkl578 these types are compatible with existing ones, and as we now support PHP 7.2, then subclasses and implementors do not break if we add a type in a superclass or abstract type.

**Comment by Majkl578** [comment]
> subclasses and implementors do not break if we add a type in a superclass or abstract type

That is only true for parameter types, but return types were also added here.

**Comment by Ocramius** [comment]
@lcobucci worth addid `UPGRADE.md` docs then

**Comment by lcobucci** [comment]
@Ocramius @Majkl578 yeap, I completely forgot to do it... sending a PR in a bit

**Pull request merged**